Delete index manifest

jwakre opened the pull request:
This PR deletes the index manifest when transitioning from format 28 -> 29. Since this manifest is directly injected into the mix, it cannot follow the typical bundle deletion workflow. As a result, there are some cases where the deletion will not be smooth:

1. Upstream mixes must create an empty bundle chroot during a format bump from 28 -> 29 for the os-core-update-index manifest to delete it.
2. Downstream mixes ahead of upstream Clear's format (format 28) will miss the index manifest deletion.
3. Downstream mixes that do not upgrade Mixer to include this change when advancing their mix to format 29 will miss the index manifest deletion.

If a mix does not successfully delete the index manifest, the following will happen:
1. The os-core-update-index manifest will be removed from the MoM.
2. The os-core-update-index manifest will be removed without deleting any of its files. Any systems that subscribed to this bundle will need to manually remove its files from their system or perform a `swupd verify --fix --picky`

jwakre commented:
@rchiossi This PR is still WIP and largely untested, but I am open to early feedback to close this PR more quickly. So far I have done the following to smoke test the PR:

```
mixer init
mixer build all --native --increment
mixer build format-bump --native --new-format 29
```

In this example, the index manifest is generated in the first mix, deleted in the +10, and does not exist in the +20.

jwakre commented on builder/bundles.go:
```diff
@@ -520,13 +520,22 @@ func installBundleToFull(packagerCmd []string, buildVersionDir string, bundle *b
+	// Index manifest content is only generated for formats < 29 and when there
+	// is no bundle chroot.
+	_, err = os.Stat(filepath.Join(buildVersionDir, "os-core-update-index"))
+	if format <= "28" && os.IsNotExist(err) {
```
The content for the os-core-update-index manifest is created during `mixer build bundles`. Should format switching in the builder package be centralized like swupd?

jwakre replied:
At a glance, I think I should be able to consolidate this in one place.

The index manifest entries are generated from the same object that populates the bundle-info files. During `mixer build update`, the bundle-info files are loaded into the `BundleInfo` object. This could be used to generate the index manifest entries.

jwakre commented on mixer/cmd/build.go:
```diff
@@ -280,6 +280,14 @@ var buildFormatOldCmd = &cobra.Command{
 		b.State.Mix.Format = buildFlags.newFormat
 
+		// Delete index manifest when transitioning from format 28 -> 29
+		if oldFormat == "28" {
```
This can be more robust: `oldFormat <= 28 && b.State.Mix.Format > 28`

rchiossi commented on swupd/manifest.go:
```diff
@@ -853,6 +856,9 @@ func writeIndexManifest(c *config, ui *UpdateInfo, bundles []*Manifest) (*Manife
 	// linkPeersAndChange will update file versions correctly
 	_, _, _ = idxMan.linkPeersAndChange(oldM, ui.minVersion)
+
+	idxMan.Header.FileCount = uint32(len(idxMan.Files))
```
Why do you need to set this again here? Does it get overwritten after line 819?

jwakre replied:
I intended to set the file count on 860 instead of 819, so I will remove the reference on 819.

When a file is deleted `linkPeersAndChange` will add a deleted file entry, but it doesn't update `idxMan.Header.FileCount`, so this value would be incorrect when deleting bundles (it leads to a failure when deleting the index manifest). Also, in the typical, bundle workflow, `idxMan.Header.FileCount` is set after `linkPeersAndChange`

jwakre commented on swupd/manifest.go:
```diff
@@ -826,23 +821,31 @@ func writeIndexManifest(c *config, ui *UpdateInfo, bundles []*Manifest) (*Manife
+	// When there is no bundle chroot for the index manifest, create a tracking file and add
+	// the allbundles directory to the index manifest. The index manifest will be deleted when
+	// a bundle chroot directory exists.
+	if _, err := os.Stat(filepath.Join(bundleDir, IndexBundle)); os.IsNotExist(err) {
```
When a bundle chroot for the index manifest exists (even if there is content in the bundle chroot), the index manifest will be deleted. Currently, index manifests do not support adding files with bundle chroots, so no functionality is lost. I do not want to enable bundle chroots for the index manifest since it is deprecated in future formats.

jwakre commented:
The PR has been updated with the following changes:
* Consolidate index manifest generation to the swupd package.
  * Previously, index manifest content was generated and injected into the full chroot during `mixer build bundles`. Now this is performed in format_switch.go.
* Added unit test to validate proper index manifest generation/deletion behavior.
* Updated index manifest code so that it will be deleted when an index manifest bundle chroot exists.

jwakre commented:
@rchiossi This PR is ready for review again.

@gtkramer When this PR is merged, Mixer will require **upstream mixes to create an empty bundle chroot for "os-core-update-index"** when performing **format bump 28 -> 29** to properly delete the bundle.

jwakre commented:
@rchiossi I am double checking that DevOps can handle the manual format-bump intervention required for this change. Please hold on merging until they sign off

jwakre commented:
> @rchiossi I am double checking that DevOps can handle the manual format-bump intervention required for this change. Please hold on merging until they sign off

DevOps has accepted the requirements to delete the index manifest for an upstream mix, so this PR is unblocked.

rchiossi commented on swupd/manifest.go:
```diff
@@ -236,6 +236,12 @@ func ParseManifestFile(path string) (*Manifest, error) {
+	if strings.Contains(m.Name, ".I.") {
+		m.Type = ManifestIterative
+	}
+
```
Unrelated to this patch, but we should probably set the types for everything that we parse. It seems like a bug that fact that this is left unset.

rchiossi commented on mixer/cmd/build.go:
```diff
@@ -280,6 +280,14 @@ var buildFormatOldCmd = &cobra.Command{
 		b.State.Mix.Format = buildFlags.newFormat
 
+		// Delete index manifest when passing format 28
+		if oldFormat <= "28" && b.State.Mix.Format > "28" {
```
This passed by me unnoticed at first, but you are comparing strings, not ints. You need to convert or this won't work as intended.

rchiossi commented on swupd/manifest.go:
```diff
@@ -882,35 +885,39 @@ func writeIndexManifest(c *config, ui *UpdateInfo, bundles []*Manifest) (*Manife
-		// in this version
-		return oldM, nil
-	}
+	if idxMan.Header.FileCount > 0 {
```
Since you are at it, this will be cleaner if you just return if FileCount <= 0 instead of moving all the code inside a block.

jwakre commented:
@rchiossi Thanks for the feedback. I rebased this PR and made the following changes to address your review:
* ParseManifest function now sets all manifest types
* Format strings are converted to ints before comparison
* When FileCount == 0, writeIndexManifest will return early
